file_transfer_agent.go

```diff
@@ -0,0 +1,687 @@
+	storageClient := sfa.getStorageClient(sfa.stageLocationType)
+
+	for _, meta := range smallFileMetadata {
+		meta.client = storageClient
+	}
+	for _, meta := range largeFileMetadata {
+		meta.client = storageClient
+	}
+
+	if len(smallFileMetadata) > 0 {
+		sfa.uploadFilesParallel(smallFileMetadata)
+	}
+	if len(largeFileMetadata) > 0 {
+		sfa.uploadFilesSequential(largeFileMetadata)
+	}
+}
+
+func (sfa *snowflakeFileTransferAgent) download(largeFileMetadata []fileMetadata, smallFileMetadata []fileMetadata) {
+	// TODO
+}
+
+func (sfa *snowflakeFileTransferAgent) uploadFilesParallel(fileMetas []*fileMetadata) {
+	idx := 0
+	fileMetaLen := len(fileMetas)
+	for idx < fileMetaLen {
+		endOfIdx := intMin(fileMetaLen, idx+int(sfa.parallel))
+		targetMeta := fileMetas[idx:endOfIdx]
+		for {
+			var wg sync.WaitGroup
+			results := make([]*fileMetadata, fileMetaLen)
+			for i, meta := range targetMeta {
+				wg.Add(1)
+				go func(k int, m *fileMetadata) {
+					defer wg.Done()
+					results[k] = sfa.uploadOneFile(m)
+				}(i, meta)
+			}
+			wg.Wait()
+
+			retryMeta := make([]*fileMetadata, 0)
+			for _, result := range results {
+				if result.resStatus == renewToken || result.resStatus == renewPresignedURL {
+					retryMeta = append(retryMeta, result)
+				} else {
+					sfa.results = append(sfa.results, result)
+				}
+			}
+
+			if len(retryMeta) == 0 {
+				break
+			}
+			for _, result := range retryMeta {
+				if result.resStatus == renewPresignedURL {
+					break
+				}
+			}
+			targetMeta = retryMeta
+		}
+		if endOfIdx == fileMetaLen {
+			break
+		}
+		idx += int(sfa.parallel)
+	}
+}
+
+func (sfa *snowflakeFileTransferAgent) uploadFilesSequential(fileMetas []*fileMetadata) {
+	idx := 0
+	fileMetaLen := len(fileMetas)
+	for idx < fileMetaLen {
+		res := sfa.uploadOneFile(fileMetas[idx])
+		sfa.results = append(sfa.results, res)
+		idx++
+		if injectWaitPut > 0 {
+			time.Sleep(injectWaitPut)
+		}
+	}
+}
+
+func (sfa *snowflakeFileTransferAgent) uploadOneFile(meta *fileMetadata) *fileMetadata {
+	fmt.Println("upload one file")
+	meta.realSrcFileName = meta.srcFileName
+	tmpDir := os.TempDir()
+	meta.tmpDir = tmpDir
+	defer os.RemoveAll(tmpDir) // cleanup
+
+	fileUtil := new(snowflakeFileUtil)
+	if meta.requireCompress {
+		if meta.srcStream != nil {
+			meta.realSrcStream, _ = fileUtil.compressFileWithGzipFromStream(&meta.srcStream)
+		} else {
+			panic("not implemented") // TODO
+		}
+	}
+
+	if meta.srcStream != nil {
+		if meta.realSrcStream != nil {
+			meta.sha256Digest, meta.uploadSize = fileUtil.getDigestAndSizeForStream(&meta.realSrcStream)
+		} else {
+			meta.sha256Digest, meta.uploadSize = fileUtil.getDigestAndSizeForStream(&meta.srcStream)
+		}
+
+	} else {
+		panic("not implemented") // TODO
+	}
+
+	client := sfa.getStorageClient(sfa.stageLocationType)
+	client.uploadOneFileWithRetry(meta)
+	return meta
+}
+
+func (sfa *snowflakeFileTransferAgent) downloadFilesParallel(fileMetas []fileMetadata) {
+	// TODO
```

Comment:
I believe this means you have not implemented GET command yet. Can you please file a JIRA and put the links (which is accessible internal only, though)?